lib/puppet/face/config.rb

```diff
@@ -159,6 +159,12 @@ def report_section_and_environment(section_name, environment_name)
         report_section_and_environment(options[:section], Puppet.settings[:environment])
       end
 
+      # Validate setting if it has a hook
+      setting = Puppet.settings.setting(name.to_sym)
+      if setting && setting.has_hook?
```

Comment:
Does this work if the setting is defined in a section other than `:main` or `:user`, but the `config` application is running in `:user` run mode? I assume yes, because we're asking for the setting directly instead of calling `Settings#value_sets_for` based on the current run mode (which would exclude the `:agent`, `:server`, and any environment section.

Reply:
From what I noticed, most of the Bad Stuff happens if we want to `get` the value of a setting. Here as we're trying to set it, we already know the value it will have, so the run mode shouldn't matter.

When getting a value, like `puppet config print serverport --section agent`, again, only the `:main` and `:user` hooks are honored. I would expect the hooks from the section we're passing to the CLI to be also evaluated.

Reply:
Since the value that we're setting could be an uninterpolated value, I think the only thing we can do is set the setting and then attempt to look it up. The hook for `certname` will get triggered when the value is set (since it's now `on_initialize_and_write`).